modules/javafx.graphics/src/main/java/com/sun/marlin/OffHeapArray.java

```diff
@@ -26,82 +26,164 @@
+    private final int instance;
+    // KCR: END DEBUG
 
     OffHeapArray(final Object parent, final long len) {
+        // KCR: BEGIN DEBUG
+        instance = instanceCount.incrementAndGet();
+        if (LOG_UNSAFE_MALLOC) {
+            System.err.println("OffHeapArray::<init> : " +
+                    "instance = " + instance +
+                    ", thread = " + Thread.currentThread());
+        }
+        // KCR: END DEBUG
+
+        arena = Arena.ofShared();
+
         // note: may throw OOME:
-        this.address = UNSAFE.allocateMemory(len);
+        // KCR FIXME: Set a MemoryLayout
+        this.segment = arena.allocate(len);
         this.length  = len;
         this.used    = 0;
         if (LOG_UNSAFE_MALLOC) {
             MarlinUtils.logInfo(System.currentTimeMillis()
                                 + ": OffHeapArray.allocateMemory =   "
-                                + len + " to addr = " + this.address);
+                                + len + " for segment = " + this.segment);
         }
 
         // Register a cleaning function to ensure freeing off-heap memory:
         MarlinUtils.getCleaner().register(parent, this::free);
     }
 
+    /**
+     * Gets the length of this array.
+     *
+     * @return the length in bytes
+     */
+    long getLength() {
+        return length;
+    }
+
+    /**
+     * Gets the number of bytes currently being used. Always <= length
+     * @return number of used bytes
+     */
+    int getUsed() {
+        return used;
+    }
+
+    /**
+     * Sets the number of bytes currently being used. Always <= length
+     * @param used number of used bytes
+     */
+    void setUsed(int used) {
+        this.used = used;
+    }
+
+    /**
+     * Increments the number of bytes currently being used.
+     * Curr used + incr used must be <= length
+     * @param used number of used bytes to increment
+     */
+    void incrementUsed(int used) {
+        this.used += used;
+    }
+
     /*
      * As realloc may change the address, updating address is MANDATORY
      * @param len new array length
      * @throws OutOfMemoryError if the allocation is refused by the system
      */
     void resize(final long len) {
-        // note: may throw OOME:
-        this.address = UNSAFE.reallocateMemory(address, len);
+        // KCR: BEGIN DEBUG
+        if (LOG_UNSAFE_MALLOC) {
+            System.err.println("OffHeapArray::resize : instance = " + instance + " len = " + len +
+                    " [was: " + this.length + ", used = " + used + "]");
+        }
+        // KCR: END DEBUG
+
+        Arena newArena = Arena.ofShared();
+        MemorySegment newSegment = newArena.allocate(len);
+
+        // KCR: Double-check this
+        // If there are any bytes in use, copy them to the newly reallocated array
+        if (this.used > 0) {
+            MemorySegment.copy(segment, 0, newSegment, 0, Math.min(this.used, len));
+        }
```

Comment:
@bourgesl I presume that there is never a need to copy more than `used` bytes?

Reply:
True.

Reply:
Question: can the number of bytes to be copied depend on memory layout (BE vs LE)?

Reply:
No